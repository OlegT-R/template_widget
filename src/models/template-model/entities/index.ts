export * from './node';
export * from './template';
export * from './textarea';
export * from './variables';
