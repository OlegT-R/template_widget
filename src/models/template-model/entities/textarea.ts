export interface ITextArea {
    id: string;
    value: string;
}
