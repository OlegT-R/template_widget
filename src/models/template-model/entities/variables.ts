
export type VariableNamesType = string[];

export interface IVariables {
  [key: string]: string,
}
