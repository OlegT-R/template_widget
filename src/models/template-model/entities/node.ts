import { ITextArea } from './textarea';

export interface IIfElse {
  id: string,
  if: NodeType;
  then: NodeType;
  else: NodeType;
}

export type NodeType = Array<ITextArea | IIfElse>;
