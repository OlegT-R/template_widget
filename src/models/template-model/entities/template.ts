import { IIfElse, NodeType } from './node';
import { ITextArea } from './textarea';

export interface ITemplate {
  block: NodeType;
}

export type NodeItem = ITextArea | IIfElse;

export function isTextArea(node: NodeItem): node is ITextArea {
  return (node as ITextArea).value !== undefined;
}

export function isIfElse(node: NodeItem): node is IIfElse {
  return (node as IIfElse).if !== undefined;
}
