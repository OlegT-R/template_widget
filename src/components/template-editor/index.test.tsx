import React from 'react';

import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import { TemplateEditor } from '.';
import { ITemplate } from '../../models/template-model/entities';

const initTemplate: ITemplate = {
  block: [
    {
      id: 'Block1Start',
      value: 'Block1Start',
    },
    {
      id: 'IfElse1',
      if: [
        {
          id: 'Block2If',
          value: 'Block2If',
        },
      ],
      then: [
        {
          id: 'BLock3Start',
          value: 'BLock3Start',
        },
        {
          id: 'IfElse2',
          if: [
            {
              id: 'BLock4If',
              value: 'BLock4If',
            },
          ],
          then: [
            {
              id: 'BLock4Then',
              value: 'BLock4Then',
            },
          ],
          else: [
            {
              id: 'BLock4Else',
              value: 'BLock4Else',
            },
          ],
        },
        {
          id: 'BLock3End',
          value: 'BLock3End',
        },
      ],
      else: [
        {
          id: 'Block2Else',
          value: 'Block2Else',
        },
      ],
    },
    {
      id: 'Block1End',
      value: 'Block1End',
    },
  ],
};

const variables = ['var1', 'var2', 'var3'];

describe('<TemplateEditor />', () => {
  test('Test Click Variable', () => {
    const saveTemplate = jest.fn();
    const component: any = mount(
      <TemplateEditor
        template={initTemplate}
        arrVarNames={variables}
        callbackSave={saveTemplate}
        showPreview={() => null}
      />
    );
    const btns = component.find('button.ant-btn-primary');
    expect(btns.length).toBe(4);
    const btnVar1 = btns.first();
    btnVar1.simulate('click');
    component.update();
    const area = component.find('textarea').first();
    const { value } = area.props();
    expect(value).toBe('Block1Start{var1}');
    expect(toJson(area)).toMatchSnapshot();
  });

  test('Test Remove 1-st level if else block', () => {
    const saveTemplate = jest.fn();
    const component: any = mount(
      <TemplateEditor
        template={initTemplate}
        arrVarNames={variables}
        callbackSave={saveTemplate}
        showPreview={() => null}
      />
    );
    const btns = component.find('button.ant-btn-danger');
    expect(btns.length).toBe(2);
    const btnDel1 = btns.first();
    btnDel1.simulate('click');
    const areas = component.find('textarea');
    expect(areas.length).toBe(1);
  });

  test('Test Adding if else block', () => {
    const saveTemplate = jest.fn();
    const component: any = mount(
      <TemplateEditor
        template={initTemplate}
        arrVarNames={variables}
        callbackSave={saveTemplate}
        showPreview={() => null}
      />
    );
    const areas = component.find('textarea');
    expect(areas.length).toBe(9);

    const canSplitArea = areas.at(1);
    canSplitArea.simulate('blur');
    component.update();
    const btnAdd = component.find('button.ant-btn-dashed');
    btnAdd.simulate('click');
    component.update();
    const newAreas = component.find('textarea');
    expect(newAreas.length).toBe(9 + 4);
  });

  test('Test call callback', () => {
    const saveTemplate = jest.fn();
    const component: any = mount(
      <TemplateEditor
        template={initTemplate}
        arrVarNames={variables}
        callbackSave={saveTemplate}
        showPreview={() => null}
      />
    );
    const btns = component.find('button.ant-btn-primary');
    btns.last().simulate('click');
    expect(saveTemplate).toBeCalled();
  });
});
