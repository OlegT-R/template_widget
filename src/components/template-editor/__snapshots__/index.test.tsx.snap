// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TemplateEditor /> Test Click Variable 1`] = `
<textarea
  className="ant-input"
  onBlur={[Function]}
  onChange={[Function]}
  onKeyDown={[Function]}
  style={
    Object {
      "height": 0,
      "maxHeight": 9007199254740991,
      "minHeight": -8,
      "overflowY": undefined,
    }
  }
  value="Block1Start{var1}"
/>
`;
